### ISSUE TITLE
Gimli doesn't support iOS

### ISSUE THREAD

**Issue by alexcrichton** [rust-lang/backtrace-rs#350]
I'm not really sure if libbacktrace supports this and if so how well. This is at least a placeholder to audit what happens tody with libbacktrace and switch gimli over to that.

**Comment by luser** [comment]
TL;DR: I don't think libbacktrace works on iOS currently, but adding iOS support to this crate shouldn't be a huge undertaking.

AFAICT libbacktrace relies on [`_Unwind_Backtrace`](https://github.com/ianlancetaylor/libbacktrace/blob/bd02146d9045f0df74a1427dbf6e45db90a5d4f1/simple.c#L106) for unwinding, which doesn't exist on iOS so I don't think libbacktrace works on iOS. Helpfully, the iOS ABIs for both [AArch64](https://developer.apple.com/library/archive/documentation/Xcode/Conceptual/iPhoneOSABIReference/Articles/ARM64FunctionCallingConventions.html#//apple_ref/doc/uid/TP40013702-SW1) as well as [ARMv7](https://developer.apple.com/library/archive/documentation/Xcode/Conceptual/iPhoneOSABIReference/Articles/ARMv7FunctionCallingConventions.html#//apple_ref/doc/uid/TP40009022-SW1) mandate the use of a frame pointer, so unwinding is straightforward in practice. [Apple's libc implementation of `backtrace`](https://opensource.apple.com/source/Libc/Libc-1272.200.26/gen/thread_stack_pcs.c.auto.html) is pretty simple: 
* Call `pthread_self` to get the current thread
* Call `pthread_get_stackaddr_np` and `pthread_get_stacksize_np` on the current thread to get the stack bounds
* Invoke `__builtin_frame_address` to get the current frame pointer
* Dereference the frame pointer in a loop to unwind the stack

I'd expect all of the existing Mach-O support in this crate to work fine for symbolication, but in practice backtraces from programs running on iOS devices won't have access to symbol files so it's immaterial. The standard practice for dealing with stack traces from iOS devices is [to symbolicate them after the fact](https://developer.apple.com/library/archive/technotes/tn2151/_index.html#//apple_ref/doc/uid/DTS40008184-CH1-SYMBOLICATEWITHXCODE). It seems like this could be handled in tooling such as [`dinghy`](https://github.com/snipsco/dinghy), although there's already [an existing issue there about backtraces not working on iOS](https://github.com/snipsco/dinghy/issues/72).

**Comment by lzutao** [comment]
Fixed by #378 ?

**Comment by cutsoy** [comment]
The current state on iOS: #378 restores all functionality that libbacktrace had, but it doesn't show line-numbers or source files. I don't know if that's expected or not? But I don't remember ever seeing line-numbers in backtraces with Rust on iOS, so it's not a regression.

**Comment by alexcrichton** [comment]
Ok makes sense, thanks for the info! In light of that I think I'll leave this open until we can get line numbers and filenames working on iOS

**Comment by fzyzcjy** [comment]
In production environment, there are no symbol names - otherwise it will be easier to reverse-engineering. However, as long as collect information on ios and can symbolize it afterwards, it is completely ok. (So can we do that?)

**Comment by workingjubilee** [comment]
The current report as of #613 is that we are now getting backtraces like this on visionOS:

```
backtrace[3e08695382c39b1]::backtrace::trace_unsynchronized::<<backtrace[3e08695382c39b1]::capture::Backtrace>::create::{closure#0}>
   1:        0x104b7d688 - backtrace[3e08695382c39b1]::backtrace::trace::<<backtrace[3e08695382c39b1]::capture::Backtrace>::create::{closure#0}>
   2:        0x104b9c518 - <backtrace[3e08695382c39b1]::capture::Backtrace>::create
   3:        0x104b9c490 - <backtrace[3e08695382c39b1]::capture::Backtrace>::new
   4:        0x104b2dea4 - <object[c239c684900fc29f]::ffi::Obj>::new
   5:        0x104b3a8c8 - object[c239c684900fc29f]::ffi::stream::object_fn_new::{closure#1}
```

And the implementation for all the Apple targets is virtually identical.

While we don't have adequate testing, I'm declaring this "fixed-ish". That last change will be in 0.3.72 which will be out Very Soon. I am certain there are oodles of bugs left in the support, or feature requests, or whatever, but please open them as _individual_ issues.
